### ISSUE TITLE
Lowering stablehlo.custom_call to linalg

### ISSUE THREAD

**Issue by endaiHW** [openxla/stablehlo#2680]
### Request description

I am trying to lower stablehlo to linalg, but the --stablehlo-legalize-to-linalg pass does not seem to convert stablehlo.custom_call. Is there a suitable pass available to handle this?

**Comment by sdasgup3** [comment]
You're right, the `--stablehlo-legalize-to-linalg` pass in MLIR doesn't directly handle `stablehlo.custom_call` operations. My best guess for this is because custom_call ops are inherently implementation-defined and can represent arbitrary functions or external library calls. There's no general way to convert them to standard Linalg operations without knowing their specific implemetation.

In any case, can you please let me know a bit more about your specific scenario. For example, is it always possible to identify target function via `call_target_name`?

**Comment by endaiHW** [comment]
I'm lowering StableHLO exported from AF2 to Linalg and encountered an external call to `lapack_ssyevd_ffi`. Can this be directly lowered to `func.func`, or are there additional steps needed to handle such external calls properly?
` %4:3 = stablehlo.custom_call @lapack_ssyevd_ffi(%3) {mhlo.backend_config = {mode = 86 : ui8, uplo = 76 : ui8}, operand_layouts = [dense<[2, 3, 1, 0]> : tensor<4xindex>], output_operand_aliases = [#stablehlo.output_operand_alias<output_tuple_indices = [0], operand_index = 0, operand_tuple_indices = []>], result_layouts = [dense<[2, 3, 1, 0]> : tensor<4xindex>, dense<[2, 1, 0]> : tensor<3xindex>, dense<[1, 0]> : tensor<2xindex>]} : (tensor<1x68x4x4xf32>) -> (tensor<1x68x4x4xf32>, tensor<1x68x4xf32>, tensor<1x68xi32>)`

**Comment by GleasonK** [comment]
Hmm. Once you're in StableHLO with custom_calls, lowering to another dialect isn't typical. In this case it seems that AF2 has some assumption that there is an lapack library to dispatch to for some math, not sure how this would be represented in linalg. I wonder if there's a way to export the model without the lapack dependency and use a reference impl of some sort.

**Comment by endaiHW** [comment]
Hi @GleasonK,
Apologies for the late reply! Our goal is to run AF2 through the MLIR pathway on CPU for subsequent optimization. We don’t need to compile for a platform without LAPACK support. As @sdasgup3 previously mentioned, lowering to func.call and linking LAPACK should suffice for our use case. Thanks for your help!

**Comment by GleasonK** [comment]
Is the conversion failing because it can't convert the custom calls? We could mark it as a legal op to unblock if so.

I'd argue that custom_call has better expressivity for external library calls than func does (i.e. it has operand and result layouts as a part of the op signature). Regardless of the op used, there will still need to be some additional work done to properly call the library, if you have any idea what the end-state IR should look like in your pipeline that would be helpful. I.e. if we represented this as a func.func, why is that better than a custom_call?

**Comment by endaiHW** [comment]
Hi @GleasonK,
Thank you for your detailed explanation and suggestions. To be honest, I'm not entirely sure what the optimal form of the IR should be either. Our primary goal is to lower the code to LLVM IR and ensure it can run correctly. I appreciate your assistance and let me know if there's anything else I can clarify or provide to move this forward.

Thanks again for your support!
